lib/updatehub/device.h

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright (c) 2018 O.S.Systems
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __DEVICE_H__
+#define __DEVICE_H__
+
+#include <zephyr.h>
+#include <hwinfo.h>
+
+static inline bool updatehub_get_device_identity(char *id, int id_max_len)
+{
+	int i;
+	char buf[2];
+	int buf_len = 0;
+	u8_t hwinfo_id[id_max_len];
+	size_t length;
+
+
+	length = hwinfo_get_device_id(hwinfo_id, sizeof(hwinfo_id));
+	if (length <= 0) {
+		return false;
+	}
+
+	memset(id, 0, id_max_len);
+
+	for (i = 0; i < length; i++) {
+		snprintk(buf, id_max_len, "%02x", hwinfo_id[i]);
+		buf_len = buf_len + strlen(buf);
+		strncat(id, buf, MIN(id_max_len, buf_len));
```

Comment:
The size value passed into strncat() represents the maximum # of chars to be appended.  So the logic is a bit wrong here.

We don't need to keep track of `buf_len`, because strlen(buf) is the maximum amount to be appended (2).   And the smallest amount to be appended would be the left over size available in id (because it's being added to each pass through the loop).

Recommend dropping use of `buf_len` and change `strncat()` line to something like:
`strncat(id, buf, MIN(id_max_len - strlen(id), strlen(buf)));`

Note: mentally compiled only.  You'll need to test that this works.

Reply:
> Recommend dropping use of `buf_len` and change `strncat()` line to something like:
> `strncat(id, buf, MIN(id_max_len - strlen(id), strlen(buf)));`

We should avoid having function calls in MIN() or MAX() macros because how those macros are implemented. In worst case, the functions might be called multiple times which is usually not desired.

Reply:
So we are going to put extra variables to avoid this.